README.md
# ubuntu-18.04-rancherd

Vagrant Box with Ubuntu 18.04 and rancherd

## Base image

Used base image [elegoev/ubuntu-18.04](https://app.vagrantup.com/elegoev/boxes/ubuntu-18.04)

## Directory Description

| directory | description                                          |
|-----------|------------------------------------------------------|
| inspec    | inspec test profiles with controls                   |
| packer    | packer build, provisioner and post-processor scripts |
| test      | test environment for provision & inspec development  |

## Configuration

### Vagrant Cloud

- [elegoev/ubuntu-18.04-rancherd](https://app.vagrantup.com/elegoev/boxes/ubuntu-18.04-rancherd)

### Useful Vagrant Plugins

- [vagrant-disksize](https://github.com/sprotheroe/vagrant-disksize)
- [vagrant-hosts](https://github.com/oscar-stack/vagrant-hosts)
- [vagrant-secret](https://github.com/tcnksm/vagrant-secret)
- [vagrant-vbguest](https://github.com/dotless-de/vagrant-vbguest)
- [vagrant-serverspec](https://github.com/vvchik/vagrant-serverspec)
- [vagrant-vmware-esxi](https://github.com/josenk/vagrant-vmware-esxi)

### Using rancher

#### Vagrantfile

    Vagrant.configure("2") do |config|

        $basebox_name="ubuntu-18.04-rancherd-test"
        $basebox_hostname="ubuntu-1804-rancherd-test"
        $src_image_name="elegoev/ubuntu-18.04-rancherd"
        $vb_group_name="basebox-rancherd-test"

        config.vm.define "#{$basebox_name}" do |machine|
          machine.vm.box = "#{$src_image_name}"
    
          # define guest hostname
          machine.vm.hostname = "#{$basebox_hostname}"

          machine.vm.provider "virtualbox" do |vb|
            vb.name = $basebox_name
            vb.cpus = 2
            vb.customize ["modifyvm", :id, "--memory", "4096" ]
            vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
            vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
            vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
            vb.customize ["modifyvm", :id, "--vram", 256 ]
          end
  
          # forwarding ports
          machine.vm.network "forwarded_port", id: "rancher-ssl",  auto_correct: true, protocol: "tcp", guest: 8443, host: 8443, host_ip: "127.0.0.1"

        end   

    end

#### Help

    rancherd --help
    rancher --help

#### Reset rancher admin access

    sudo rancherd reset-admin

### Referenzen

- [Introducing RancherD: A Simpler Tool for Deploying Rancher](https://rancher.com/blog/2020/rancherd-simplifies-rancher-deployment)
- [webfs homepage](https://linux.bytesex.org/misc/webfs.html)

### Versioning

Repository follows sematic versioning  [![semantic versioning](https://img.shields.io/badge/semver-2.0.0-green.svg)](http://semver.org)

### Changelog

For all notable changes see [CHANGELOG](https://github.com/elegoev/basebox-ubuntu-18.04-rancherd/blob/master/CHANGELOG.md)

### License

Licensed under The MIT License (MIT) - for the full copyright and license information, please view the [LICENSE](https://github.com/elegoev/basebox-ubuntu-18.04-rancherd/blob/master/LICENSE) file.

### Issue Reporting

Any and all feedback is welcome.  Please let me know of any issues you may find in the bug tracker on github. You can find it [here.](https://github.com/elegoev/basebox-ubuntu-18.04-rancherd/issues)


CHANGELOG.md
# Change Log

## 1.0.x

**Implemented enhancements:**

- rancher installation
- Packer build implemented


packer/provisioners/base-provisioning.sh
#!/bin/bash

application_file_path="/vagrant/installed-application.md"
rke2_kubeconfig_file="/etc/rancher/rke2/rke2.yaml"
vagrant_kubeconfig_dir="/home/vagrant/.kube"
vagrant_kubeconfig_file="config"
root_kubeconfig_dir="/root/.kube"
rancher_cli_version="2.4.10"

# install rancher client
echo ">>>>>> rancher client"
sudo wget -q -O /tmp/rancher-linux-amd64-v${rancher_cli_version}.tar.gz https://releases.rancher.com/cli2/v${rancher_cli_version}/rancher-linux-amd64-v${rancher_cli_version}.tar.gz
sudo tar -xzf /tmp/rancher-linux-amd64-v${rancher_cli_version}.tar.gz -C /opt
sudo ln /opt/rancher-v${rancher_cli_version}/rancher /usr/bin/rancher
RANCHER_VERSION=$(rancher --version | awk  '{print $3}' | tr --delete v)
echo "rancher client = ${RANCHER_VERSION}"

# install rancherd
echo ">>>>>> rancherd"
sudo curl -sfL https://get.rancher.io | sh -

# enable rancherd service
echo ">>>>>> rancherd service"
sudo systemctl enable rancherd-server.service
sudo systemctl start rancherd-server.service
sudo systemctl status rancherd-server.service
retcode=$?
if [ $retcode -ne 0 ]; then
  exit 1
fi

# check kubeconfig file
echo "Wait kubeconfig ..."
while :
do
  if [ -f "${rke2_kubeconfig_file}" ]; then
    break;
  fi
  echo "."
  sleep 5
done

# copy kubeconfig for vagrant
if [ ! -d "${vagrant_kubeconfig_dir}" ]; then
  sudo mkdir -p ${vagrant_kubeconfig_dir}
fi
sudo cp ${rke2_kubeconfig_file} ${vagrant_kubeconfig_dir}/${vagrant_kubeconfig_file}
sudo chown -R vagrant:vagrant ${vagrant_kubeconfig_dir}

# copy kubeconfig for root
sudo cp -r ${vagrant_kubeconfig_dir} ${root_kubeconfig_dir}

# add rke2 bin directory to path
echo "export PATH=$PATH:/var/lib/rancher/rke2/bin" >> /home/vagrant/.bashrc
export PATH=$PATH:/var/lib/rancher/rke2/bin

# check installation
echo "Wait kubernetes cluster ..."
sleep 180
while :
do
  sudo rancherd reset-admin > /dev/null 2>&1
  retcode=$?
  if [ $retcode -eq 0 ]; then
    break
  fi
  echo "."
  sleep 30
done
sleep 10
echo "Kubernetes is up & running!"
# kubectl get daemonset rancher -n cattle-system
# kubectl get pod -n cattle-system
sudo rancherd reset-admin

# set version
RANCHERD_VERSION=$(rancherd --version | grep rancherd | awk  '{print $3}' | tr --delete v)
RANCHER_VERSION=$(rancher --version | awk  '{print $3}' | tr --delete v)
echo "# Installed application "  > $application_file_path
echo "***                     " >> $application_file_path
echo "> RancherD: $RANCHERD_VERSION" >> $application_file_path
echo "> Rancher Client: $RANCHER_VERSION" >> $application_file_path




inspec/baseline-ubuntu-18.04-rancherd/controls/check_rancherd.rb
# copyright: 2020, Urs Voegele

title "check rancherd"

# check rancherd package
control "rancherd-1.0" do                      # A unique ID for this control
  impact 1.0                                   # The criticality, if this control fails.
  title "check if rancherd command exists"     # A human-readable title
  desc "check rancherd command"
  describe command('rancherd --version') do
    its('exit_status') { should eq 0 }
  end
end

# check rancherd service
control "rancherd-2.0" do                     # A unique ID for this control
  impact 1.0                                  # The criticality, if this control fails.
  title "check if rancherd is running"        # A human-readable title
  desc "check rancherd service"
  describe service('rancherd-server') do
    it { should be_installed }
    it { should be_enabled }
    it { should be_running }
  end
end


packer/post-processors/create-git-tag.ps1
# create git tag

$src_image_version = $env:IMAGE_VERSION

$vagrant_file_template = '
Vagrant.configure("2") do |config|

    $basebox_name="ubuntu-18.04-rancherd-test"
    $basebox_hostname="ubuntu-1804-rancherd-test"
    $src_image_name="elegoev/ubuntu-18.04-rancherd"
    $vb_group_name="basebox-rancherd-test"

    config.vm.define "#{$basebox_name}" do |machine|
      machine.vm.box = "#{$src_image_name}"
      machine.vm.box_version = "$image_version"
    
      # define guest hostname
      machine.vm.hostname = "#{$basebox_hostname}"

      machine.vm.provider "virtualbox" do |vb|
        vb.name = $basebox_name
        vb.cpus = 1
        vb.customize ["modifyvm", :id, "--memory", "1024" ]
        vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
        vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
        vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
        vb.customize ["modifyvm", :id, "--vram", 256 ]
      end
  
      # forwarding ports
      machine.vm.network "forwarded_port", id: "rancher-ssl",  auto_correct: true, protocol: "tcp", guest: 8443, host: 8443, host_ip: "127.0.0.1"

    end   

end
'

# create Vagrantfile for download test
$vagrant_file_expanded = $vagrant_file_template.Replace('$image_version', "$src_image_version")
Set-Content -Path './test/download/Vagrantfile' -Value "$vagrant_file_expanded"

# create git tag
git add --all
git commit -m "new basebox version $src_image_version"
git push 
git tag $src_image_version
git push --tags





